src/sentry/search/django/backend.py

```diff
@@ -136,12 +136,22 @@ def _build_queryset(self, project, query=None, status=None, tags=None,
             )
 
         if tags:
-            matches = self._tags_to_filter(project, tags)
-            if not matches:
-                return queryset.none()
-            queryset = queryset.filter(
-                id__in=matches,
-            )
+            filter_tags = {k: v for k, v in six.iteritems(tags) if k[0] != '-'}
+            exclude_tags = {k[1:]: v for k, v in six.iteritems(tags) if k[0] == '-'}
+            if filter_tags:
+                matches = self._tags_to_filter(project, filter_tags)
+                if not matches:
+                    return queryset.none()
+                queryset = queryset.filter(
+                    id__in=matches,
+                )
+            if exclude_tags:
+                matches = self._tags_to_filter(project, exclude_tags)
+                if not matches:
+                    return queryset.none()
```

Comment:
I don't think this logic makes sense here for a few reasons:

1. If there no matches because a tag doesn't exist, or a tag has no recent values, `_tags_to_filter` will return `None`. This mean that if there is nothing to exclude, the search will return no results.
2. `_tags_to_filter` assumes we're reducing the possible input set with each tag — for example, if the input set was `[1,2,3,4,5]` and the filters represented "all even numbers" and "numbers divisible by 4 with no remainder", the only item in the input set that matches all constraints is `4`. This isn't how exclusion filters typically work, though — if we were to *exclude* values using the exact same filter criteria, the expected result should be `[1, 3, 5]` since `2` is an even number and `4` is both an even number along with evenly divisible by `4`. The fact that `2` isn't evenly divisible by `4` isn't relevant. Instead of excluding values that contain *all* of the excluded tag values here, we'd actually want to exclude values that contain *any* of the included tag values. This way you can write filters like `-level:info -level:debug`. As it's written right now, a query of `-level:info -level:debug` would only exclude groups that have been tagged with *both* of those values.